mage_ai/data_cleaner/column_type_detector.py

```diff
@@ -54,143 +56,117 @@ def get_mismatched_row_count(series, column_type):
+    clean_series = series.apply(lambda x: x.strip(' \'\"') if type(x) is str else x)
+    clean_series = clean_series.map(
+        lambda x: x if (not isinstance(x, str) or x != '') else np.nan,
+    )
+    clean_series = clean_series.dropna()
+
+    series_nunique = series.nunique(dropna=False)
+    clean_series_nunique = clean_series.nunique()
+
+    exact_dtype = type(clean_series.iloc[0]) if clean_series.count() else None
+    if exact_dtype is list:
+        mdtype = TEXT
+    elif np.issubdtype(exact_dtype, np.bool_):
+        if clean_series_nunique <= 2:
+            mdtype = TRUE_OR_FALSE
+        else:
+            mdtype = CATEGORY
+    elif clean_series_nunique <= 2:
+        mdtype = TRUE_OR_FALSE
+    else:
+        if all(clean_series.str.match(REGEX_NUMBER)):
+            if clean_series.str.contains(REGEX_FLOAT_NEW_SYM).sum():
+                mdtype = NUMBER_WITH_DECIMALS
             else:
-                integer_feature_names.append(col_name)
-
-    number_feature_names = float_feature_names + integer_feature_names
-    binary_feature_names += \
-        [col for col in number_feature_names if df[col].nunique(dropna=False) == 2]
-    binary_feature_names += \
-        [col for col in non_number_feature_names if df[col].nunique(dropna=False) == 2]
-    float_feature_names = [col for col in float_feature_names if col not in binary_feature_names]
-    integer_feature_names = \
-        [col for col in integer_feature_names if col not in binary_feature_names]
-
-    for col_name in subtract(non_number_feature_names, binary_feature_names):
-        df_drop_na = df[col_name].dropna()
-        if df_drop_na.empty:
-            text_feature_names.append(col_name)
+                mdtype = NUMBER
         else:
-            matches = pd.to_datetime(df_drop_na, infer_datetime_format=True, errors='coerce')
-            if type(df_drop_na.iloc[0]) is list:
-                text_feature_names.append(col_name)
-            elif matches.count() / len(matches) >= DATETIME_MATCHES_THRESHOLD:
-                datetime_feature_names.append(col_name)
-            elif df_drop_na.nunique() / len(df_drop_na) >= 0.8:
-                text_feature_names.append(col_name)
+            length = len(clean_series)
+            matches = clean_series.str.match(REGEX_DATETIME).sum()
+            if matches / length >= DATETIME_MATCHES_THRESHOLD:
+                mdtype = DATETIME
             else:
-                word_count, _ = \
-                    df[col_name].dropna().map(lambda x: (len(str(x).split(' ')), str(x))).max()
-                if word_count > MAXIMUM_WORD_LENGTH_FOR_CATEGORY_FEATURES:
-                    text_feature_names.append(col_name)
+                correct_emails = clean_series.str.match(REGEX_EMAIL).sum()
+                correct_phone_nums = clean_series.str.match(REGEX_PHONE_NUMBER).sum()
+                correct_zip_codes = clean_series.str.match(REGEX_ZIP_CODE).sum()
+                if correct_emails / length >= 0.90:
```

Comment:
This threshold seems pretty high. Do you have particular reason of choosing a high threshold?

Reply:
My thoughts initially were that if a column is filled with a more niche data type such as emails and phone numbers, there should be a larger number of matches. That said, I can definitely think of many instances in which that wouldn't be the case and that this threshold could be too large (e.g. people fill in the email column incorrectly).

The original threshold was 0.01, which I feel to be too low. At this threshold, a large enough dataset with random strings might even accidentally match one of the patterns a few times (albeit quite rarely). 

How does a threshold of 0.3 sound? Should be more resistant to noise in the dataset yet still be flexible with regards to human error.

Reply:
Yes, 0.3 sounds good to me.

Reply:
Fixed threshold